src/hooks/useBalanceChartData/useBalanceChartData.ts

```diff
@@ -0,0 +1,333 @@
+    [k: CAIP19]: Asset
+  }
+  priceHistoryData: {
+    [k: CAIP19]: HistoryData[]
+  }
+}
+
+type FiatBalanceAtBucketReturn = number
+
+type FiatBalanceAtBucket = (args: FiatBalanceAtBucketArgs) => FiatBalanceAtBucketReturn
+
+const fiatBalanceAtBucket: FiatBalanceAtBucket = ({
+  bucket,
+  priceHistoryData,
+  portfolioAssets
+}) => {
+  const { balance, end } = bucket
+  // TODO(0xdef1cafe): this isn't super accurate, we can
+  // a) interpolate for more accuracy (still not much better), or
+  // b) fetch a price at each specific tx time
+  const time = end.valueOf()
+  const { crypto } = balance
+  const result = Object.entries(crypto).reduce((acc, [caip19, assetCryptoBalance]) => {
+    const assetPriceHistoryData = priceHistoryData[caip19]
+    const price = priceAtBlockTime({ assetPriceHistoryData, time })
+    const portfolioAsset = portfolioAssets[caip19]
+    const { precision } = portfolioAsset
+    const assetFiatBalance = bn(assetCryptoBalance)
+      .div(bn(10).exponentiatedBy(precision))
+      .times(price)
+      .toNumber()
+    return acc + assetFiatBalance
+  }, 0)
+  return result
+}
+
+type CalculateBucketPricesArgs = {
+  assets: CAIP19[]
+  buckets: Bucket[]
+  portfolioAssets: {
+    [k: CAIP19]: Asset
+  }
+  priceHistoryData: {
+    [k: CAIP19]: HistoryData[]
+  }
+}
+
+type CalculateBucketPrices = (args: CalculateBucketPricesArgs) => Bucket[]
+
+// note - this mutates buckets
+const calculateBucketPrices: CalculateBucketPrices = (args): Bucket[] => {
+  const { assets, buckets, portfolioAssets, priceHistoryData } = args
+  // we iterate from latest to oldest
+  for (let i = buckets.length - 1; i >= 0; i--) {
+    const bucket = buckets[i]
+    const { txs } = bucket
+
+    // copy the balance back from the most recent bucket
+    bucket.balance = Object.assign(
+      {},
+      buckets[i + 1]?.balance || buckets[buckets.length - 1].balance
+    )
+
+    // if we have txs in this bucket, adjust the crypto balance in each bucket
+    txs.forEach(tx => {
+      const assetCAIP19 = caip19FromTx(tx)
+      // don't calculate price for this asset because we didn't ask for it
+      if (!assets.includes(assetCAIP19)) return
+
+      const { type, value: valueString } = tx
+      const feeValue = bnOrZero(tx.fee?.value)
+      const value = bnOrZero(valueString) // tx value in base units
+      switch (type) {
+        case TxType.Send: {
+          const amount = value.plus(feeValue)
+          const cryptoDiff = bn(amount).toNumber()
+          bucket.balance.crypto[assetCAIP19] += cryptoDiff // we're going backwards, so a send means we had more before
+          break
+        }
+        case TxType.Receive: {
+          const cryptoDiff = bn(value).toNumber()
+          bucket.balance.crypto[assetCAIP19] -= cryptoDiff // we're going backwards, so a receive means we had less before
+          break
+        }
+        default: {
+          throw new Error(`useBalanceChartData: invalid tx.type ${type}`)
```

Comment:
continue rather than throw, but this should never happen with TxType only being Send or Receive